### PR TITLE
aws: reintroduce SNP-based attestation

### PR DESCRIPTION
**Pull request by derpsteb**
### Context

As far as we can tell AWS has now fixed all issues regarding booting SNP machines and generating attestation reports. 
Thus we want to re enable SNP-based attestation to Constellation. Since multiple aspects changed (attesationconfigapi, verify cmd, go-sev-guest for validation) since the original implementation of AWS SNP attestation, this PR is a superset of the original implementation in https://github.com/edgelesssys/constellation/pull/1916.

This PR refactors the verify command, the attestationconfigapi client & fetcher and the attestationconfigapi CLI in order to not introduce duplicate code.

### Proposed change(s)

- refactor verify to be more CSP agnostic. 
- add the actual attestation implementation. The default TCB values are hardcoded and replaced in a later commit.
- adjusts the verify command to fetch and handle VLEK certificates. Without this, VLEKs are not displayed correctly (cspid vs hwid) and no certificate chain is fetched.
- refactor `internal/api/attestationconfigapi` to receive a variant argument. Previously the whole code was built for Azure. By introducing a variant argument we can reuse the code for AWS. 
- Add support to attestationconfigapi CLI for a csp and kind argument. The CSP argument is used to select whether to upload an snp-report for azure or aws. The kind argument will be used to upload launchmeasurement metadata. This was planned for this PR originally, which is why it is included here. However, we decided to leave it out for now as some open points need to be discussed.
- Adjust CI to upload values for AWS 

### Checklist

- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [ ] Link to Milestone

### CODE REVIEW

File: internal/attestation/aws/snp/validator.go [hunk 1]
```diff
@@ -9,99 +9,210 @@ package snp
...
 		log,
 	)
-	v.getDescribeClient = getEC2Client
 	return v
 }
 
 // getTrustedKeys return the public area of the provides attestation key.
-// Normally, here the trust of this key should be verified, but currently AWS does not provide this feature.
-func getTrustedKey(_ context.Context, attDoc vtpm.AttestationDocument, _ []byte) (crypto.PublicKey, error) {
+// Normally, the key should be verified here, but currently AWS does not provide means to do so.
```
Comment (daniel-weisse) [inline internal/attestation/aws/snp/validator.go:70]: Are we not validating this key now using the SNP report?

File: internal/attestation/aws/snp/validator.go [hunk 2]
```diff
@@ -9,99 +9,210 @@ package snp
...
-	if len(imageOutput.Images) > 1 {
-		return fmt.Errorf("found multiple image references for image ID %s", imageID)
+
+	validateOpts := &validate.Options{
+		// Check that the attestation key's digest is included in the report.
+		ReportData: akDigest[:],
+		GuestPolicy: abi.SnpPolicy{
+			Debug: false, // Debug means the VM can be decrypted by the host for debugging purposes and thus is not allowed.
+			SMT:   true,  // Allow Simultaneous Multi-Threading (SMT). Normally, we would want to disable SMT
+			// but Azure does not allow to disable it.
```
Comment (daniel-weisse) [inline internal/attestation/aws/snp/validator.go:167]: This is for AWS, is SMT disabled here?

**Review by daniel-weisse — COMMENTED** [review]
Didn't look at any of the CLI stuff yet, and didnt do an indepth review of the attestation parts

File: internal/api/attestationconfigapi/cli/main.go [hunk 3]
```diff
@@ -56,29 +59,40 @@ func main() {
...
-		Long: fmt.Sprintf("The CLI uploads an observed version number specific to the azure-sev-snp attestation variant to a cache directory. The CLI then determines the lowest version within the cache-window present in the cache and writes that value to the config api if necessary. "+
+	rootCmd.AddCommand(newUploadCmd())
+	rootCmd.AddCommand(newDeleteCmd())
+
+	return rootCmd
+}
+
+func newUploadCmd() *cobra.Command {
+	uploadCmd := &cobra.Command{
+		Use:   "upload {azure|aws} {snp-report|guest-firmware} <path>",
```
[the remark this answers is not in the document]
Comment (derpsteb) [reply internal/api/attestationconfigapi/cli/main.go:75]: I have changed the interface here because the path is an argument for the action you want to take. Not an option (flag) that modifies the action. I also find it less confusing than having flags. I would ask others for their opinion. @elchead @msanft wdyt?

File: internal/attestation/aws/snp/validator.go (hunk 2 above)
Comment (derpsteb) [reply internal/attestation/aws/snp/validator.go:167]: No, because they told us not to. I asked them about an update and fixed the comment.

File: internal/attestation/aws/snp/validator.go (hunk 1 above)
Comment (derpsteb) [reply internal/attestation/aws/snp/validator.go:70]: updated the comment to be more precise.

File: internal/api/attestationconfigapi/cli/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (elchead) [reply internal/api/attestationconfigapi/cli/main.go]: While I see that deeply nested commands are uncommon, I think that it reads better as is atm. Since it's an internal tool I have no strong opinion and would honestly just keep as is.

File: internal/api/attestationconfigapi/cli/e2e/test.sh.in
Comment (daniel-weisse) [inline internal/api/attestationconfigapi/cli/e2e/test.sh.in:1]: Add tests for aws? (Or make the test configurable to run either azure or aws)

File: internal/attestation/aws/snp/validator.go [hunk 4]
```diff
@@ -9,99 +9,212 @@ package snp
...
 
-// getTrustedKeys return the public area of the provides attestation key.
-// Normally, here the trust of this key should be verified, but currently AWS does not provide this feature.
-func getTrustedKey(_ context.Context, attDoc vtpm.AttestationDocument, _ []byte) (crypto.PublicKey, error) {
+// getTrustedKeys return the public area of the provided attestation key (AK).
+// Ideally, the AK should be bound to the TPM via an endorsement key, but currently AWS does not provide one.
+// The AK's digest is written to the SNP report's userdata field during report generation.
+// The AK is trusted if the report can be verified and the AK's digest matches the digest of the AK in attDoc.
+func (v *Validator) getTrustedKey(ctx context.Context, attDoc vtpm.AttestationDocument, _ []byte) (crypto.PublicKey, error) {
 	// Copied from https://github.com/edgelesssys/constellation/blob/main/internal/attestation/qemu/validator.go
```
Comment (daniel-weisse) [inline internal/attestation/aws/snp/validator.go:74]:
```suggestion
```

File: internal/attestation/aws/snp/validator.go [hunk 5]
```diff
@@ -9,99 +9,212 @@ package snp
...
-		return err
+		return [64]byte{}, fmt.Errorf("marshalling public key: %w", err)
+	}
+
+	return sha512.Sum512(pub), nil
+}
+
+// Validate a given SNP report.
+type snpReportValidator interface {
+	validate(ctx context.Context, attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, ak [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error
```
Comment (daniel-weisse) [inline internal/attestation/aws/snp/validator.go:109]:
```suggestion
	validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, ak [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error
```
Since AWS and Azure don't share a common interface, you can drop the context from this method as its not used by the AWS code

File: internal/attestation/snp/snp_test.go [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+				assert.Error(err)
+			} else {
+				assert.NoError(err)
+				assert.Equal(tc.wantVCEK, vcek != nil)
+			}
+		})
+	}
+}
+
+const (
```
Comment (daniel-weisse) [inline internal/attestation/snp/snp_test.go:128]: We have most certs saved in the `testdata` folder, why not these?

File: internal/attestation/snp/snp_test.go [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+				assert.NoError(err)
+				assert.Equal(tc.wantAsk, ask != nil)
+				assert.Equal(tc.wantArk, ark != nil)
+			}
+		})
+	}
+}
+
+// TestParseVCEK tests the parsing of the VCEK certificate.
+func TestParseVCEK(t *testing.T) {
```
Comment (daniel-weisse) [inline internal/attestation/snp/snp_test.go:86]: Since the test harness is pretty much the same, you could merge this with `TestParseCertChain`

File: internal/api/attestationconfigapi/cli/delete.go [hunk 8]
```diff
@@ -23,62 +22,83 @@ import (
 // newDeleteCmd creates the delete command.
 func newDeleteCmd() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     "delete",
-		Short:   "delete a specific version from the config api",
+		Use:     "delete {azure|aws} {snp-report|guest-firmware} <version>",
```
[the remark this answers is not in the document]
Comment (elchead) [reply internal/api/attestationconfigapi/cli/delete.go:25]: On the other hand flags suggest that sth is optional and what would the command do without flags?

[the remark this answers is not in the document]
Comment (daniel-weisse) [reply internal/api/attestationconfigapi/cli/delete.go:25]: You can make flags not optional: https://pkg.go.dev/github.com/spf13/cobra#MarkFlagRequired

[the remark this answers is not in the document]
Comment (elchead) [reply internal/api/attestationconfigapi/cli/delete.go:25]: I keep it to Otto to change it or not. fine with both

File: internal/attestation/snp/snp_test.go (hunk 7 above)
Comment (derpsteb) [reply internal/attestation/snp/snp_test.go:86]: Not sure how you mean that. The functions do different things and have different return values. The testcases are also different so there would be no overlap.

Comment (daniel-weisse) [reply internal/attestation/snp/snp_test.go:86]: Ah sorry, misread the function under test

File: internal/api/attestationconfigapi/reporter.go [hunk 9]
```diff
@@ -6,6 +6,7 @@ SPDX-License-Identifier: AGPL-3.0-only
 
 /*
 The reporter contains the logic to determine a latest version for Azure SEVSNP based on cached version values observed on CVM instances.
+The code in this file does not rely on dedicated API objects for no other reason than original development speed.
```
[the remark this answers is not in the document]
Comment (derpsteb) [reply internal/api/attestationconfigapi/reporter.go:9]: Slightly adjusted comment to be more precise. Does that clarify things?

File: internal/api/attestationconfigapi/snp.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	TEE uint8 `json:"tee"`
+	// SNP is the latest version of the Azure SEVSNP SNP.
+	SNP uint8 `json:"snp"`
+	// Microcode is the latest version of the Azure SEVSNP microcode.
+	Microcode uint8 `json:"microcode"`
+}
+
+// SEVSNPVersionAPI is the request to get the version information of the specific version in the config api.
+// Because variant is not part of the marshalled JSON, fetcher and client methods need to fill the variant property.
+// Once we switch to v2 of the API we should embed the variant in the object.
```
[the remark this answers is not in the document]
Comment (derpsteb) [reply internal/api/attestationconfigapi/snp.go:36]: I am not sure I understand the question but these two facts made me write the comment:
- we can't add a property to the JSONs in the API within one version because old clients will break. this is why we can't put the variant into the JSON in API version 1.
- we want the variant in the JSON so `client.Fetch(..)` returns a SEVSNPVersionAPI object that has a valid `Variant` property set.

File: internal/api/attestationconfigapi/snp.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+// That would remove the possibility of some fetcher/client code forgetting to set the variant.
+type SEVSNPVersionAPI struct {
```
[the remark this answers is not in the document]
Comment (derpsteb) [reply internal/api/attestationconfigapi/snp.go:38]: This would not prevent a new developer from using the object that is returned by `client.Fetch(someSEVNSNPVersionAPI)`. We can't put a constructor into the Fetch code as it would break generality.

File: internal/api/attestationconfigapi/fetcher.go [hunk 12]
```diff
@@ -64,35 +65,45 @@ func newFetcherWithClientAndVerifier(client apifetcher.HTTPClient, cosignVerifie
 	return &fetcher{HTTPClient: client, verifier: cosignVerifier, cdnURL: url}
 }
 
-// FetchAzureSEVSNPVersionList fetches the version list information from the config API.
-func (f *fetcher) FetchAzureSEVSNPVersionList(ctx context.Context, attestation AzureSEVSNPVersionList) (AzureSEVSNPVersionList, error) {
+// FetchSEVSNPVersionList fetches the version list information from the config API.
+func (f *fetcher) FetchSEVSNPVersionList(ctx context.Context, list SEVSNPVersionList) (SEVSNPVersionList, error) {
 	// TODO(derpsteb): Replace with FetchAndVerify once we move to v2 of the config API.
```
[the remark this answers is not in the document]
Comment (derpsteb) [reply internal/api/attestationconfigapi/fetcher.go:70]: Verifying `list.sig` would require that we consistently create the signature with all operations that modify the list of versions. Since v1 does not dictate the existence of a `list.sig` file there might be clients that update `list` without `list.sig`. Those clients are valid for v1. That update from a v1 client in turn would break clients that verify `list.sig`.

[the remark this answers is not in the document]
Comment (elchead) [reply internal/api/attestationconfigapi/fetcher.go:70]: good catch

**Comment by derpsteb** [comment]
Regarding the CLI UX: since Daniel said he would be fine with the current proposal if the CLI is only used in the CI, which it is, I would keep things as is for now. If Moritz or Thomas don't raise more concerns.

File: internal/attestation/aws/snp/issuer.go [hunk 13]
```diff
@@ -49,18 +54,69 @@ func getAttestationKey(tpm io.ReadWriter) (*tpmclient.Key, error) {
...
-	return func(ctx context.Context, _ io.ReadWriteCloser, _ []byte) ([]byte, error) {
-		ec2InstanceIdentityOutput, err := client.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
-		if err != nil {
-			return nil, fmt.Errorf("fetching instance identity document: %w", err)
-		}
-		return json.Marshal(ec2InstanceIdentityOutput.InstanceIdentityDocument)
+func getInstanceInfo(_ context.Context, tpm io.ReadWriteCloser, _ []byte) ([]byte, error) {
+	tpmAk, err := client.AttestationKeyRSA(tpm)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RSA Endorsement key: %w", err)
```
Comment (msanft) [inline internal/attestation/aws/snp/issuer.go:63]:
```suggestion
		return nil, fmt.Errorf("creating RSA Endorsement key: %w", err)
```
nit

File: internal/attestation/aws/snp/validator.go [hunk 14]
```diff
@@ -9,99 +9,211 @@ package snp
...
-		return err
+		return [64]byte{}, fmt.Errorf("marshalling public key: %w", err)
+	}
+
+	return sha512.Sum512(pub), nil
+}
+
+// Validate a given SNP report.
+type snpReportValidator interface {
+	validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, ak [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error
```
Comment (msanft) [inline internal/attestation/aws/snp/validator.go:108]:
```suggestion
// snpReportValidator validates a given SNP report.
type snpReportValidator interface {
	validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, ak [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error
```
nit

File: internal/attestation/aws/snp/validator.go [hunk 15]
```diff
@@ -9,99 +9,211 @@ package snp
...
+}
+
+// Validation logic for the AWS SNP implementation.
+type awsValidator struct {
+	verifier    reportVerifier
+	validator   reportValidator
+	httpsGetter trust.HTTPSGetter
+}
```
Comment (msanft) [inline internal/attestation/aws/snp/validator.go:116]:
```suggestion
// Validation logic for the AWS SNP implementation.

type awsValidator struct {
	verifier    reportVerifier
	validator   reportValidator
	httpsGetter trust.HTTPSGetter
}
```
nit: the comment seems unrelated to what `awsValidator` is specifically and rather targets all the code below.

File: internal/attestation/aws/snp/validator.go [hunk 16]
```diff
@@ -9,99 +9,211 @@ package snp
...
+type reportVerifierImpl struct{}
+
+func (r *reportVerifierImpl) SnpAttestation(att *sevsnp.Attestation, opts *verify.Options) error {
+	return verify.SnpAttestation(att, opts)
+}
+
+// validate the report by checking if it has a valid VLEK signature.
+// The certificate chain ARK -> ASK -> VLEK is also validated.
+// Checks that the report's userData matches the connection's userData.
+func (a *awsValidator) validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, akDigest [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error {
```
Comment (msanft) [inline internal/attestation/aws/snp/validator.go:140]: Note: We could possibly make this more generic at some point, as it shares a lot of logic with the Azure validator. But not in this PR.

File: internal/attestation/aws/snp/validator.go [hunk 17]
```diff
@@ -9,99 +9,211 @@ package snp
...
+	validateOpts := &validate.Options{
+		// Check that the attestation key's digest is included in the report.
+		ReportData: akDigest[:],
+		GuestPolicy: abi.SnpPolicy{
+			Debug: false, // Debug means the VM can be decrypted by the host for debugging purposes and thus is not allowed.
+			SMT:   true,  // Allow Simultaneous Multi-Threading (SMT). Normally, we would want to disable SMT
+			// but AWS machines are currently facing issues if it's disabled.
+		},
+		VMPL: new(int), // Checks that Virtual Machine Privilege Level (VMPL) is 0.
+		// This checks that the reported LaunchTCB version is equal or greater than the minimum specified in the config.
```
Comment (msanft) [inline internal/attestation/aws/snp/validator.go:171]: We could possibly leave a sentence here explaining why we only check the Launch TCB

File: internal/attestation/aws/snp/validator_test.go [hunk 18]
```diff
@@ -7,41 +7,71 @@ SPDX-License-Identifier: AGPL-3.0-only
...
+	spb "github.com/google/go-sev-guest/proto/sevsnp"
+	"github.com/google/go-sev-guest/verify"
 	"github.com/google/go-tpm-tools/proto/attest"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
-func TestGeTrustedKey(t *testing.T) {
+const (
+	// validAKDigest is the AK hash of the public key embedded in 'reportValid'.
```
Comment (msanft) [inline internal/attestation/aws/snp/validator_test.go:39]: Possibly i'm just blind, but where is `reportValid`?

File: internal/attestation/aws/snp/validator_test.go [hunk 19]
```diff
@@ -7,41 +7,71 @@ SPDX-License-Identifier: AGPL-3.0-only
...
+	validAKDigest = "032635613c8e331fa29e096371910fe6a1f69383dda02c9461400a70b66d87a3da5dd863002522be43afc34f2c233989bd6e401e351d10d7cc800d6f5dfcf019"
+)
+
+func TestGetTrustedKey(t *testing.T) {
+	validator := &Validator{ask: nil, ark: nil, reportValidator: stubawsValidator{}}
```
Comment (msanft) [inline internal/attestation/aws/snp/validator_test.go:44]:
```suggestion
	validator := func() *Validator {return &Validator{ask: nil, ark: nil, reportValidator: stubawsValidator{}}}
```
nit: prevent issues arising from shared state

File: internal/attestation/aws/snp/validator_test.go [hunk 20]
```diff
@@ -63,84 +94,219 @@ func TestGeTrustedKey(t *testing.T) {
...
+// - parse cached AK hash. Hash and SNP report have to match.
+// - parse cache VLEK cert.
+func TestValidateSNPReport(t *testing.T) {
+	require := require.New(t)
+	certs, err := loadCerts(testdata.CertChain)
+	require.NoError(err)
+	ark := certs[1]
+	certs, err = loadCerts(testdata.CertChain)
+	require.NoError(err)
+	ask := certs[0]
```
Comment (msanft) [inline internal/attestation/aws/snp/validator_test.go:110]:
```suggestion
	ark := certs[1]
	ask := certs[0]
```
Or is there a reason for this?

File: internal/attestation/snp/snp_test.go [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
...0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000b9853dac65f127574c6a578c11885e1887d4c7ae446237d4273715dd8c05cfe4bd49facc1392f2ca7354c8f0d34d65500000000000000000000000000000000000000000000000004013481e9c6a6bb112818aeba3bd178d788dedf62600b8c7892a8d3df4d880265010e7d833201156364a001e62f47b570000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
```
Comment (msanft) [inline internal/attestation/snp/snp_test.go:26]: We should move this to a testdata package

File: internal/config/azure.go [hunk 22]
```diff
@@ -17,6 +17,9 @@ import (
...
...M9eNxhIySb2npJfgnqz55I0u33wh4r0ZNQ\neTGfw03MBUtyuzGesGkcw+loqMaq1qR4tjGbPYxCvpCq7+OgpCCoMNit2uLo9M18\nfHz10lOMT8nWAUvRZFzteXCm+7PHdYPlmQwUw3LvenJ/ILXoQPHfbkH0CyPfhl1j\nWhJFZasCAwEAAaN+MHwwDgYDVR0PAQH/BAQDAgEGMB0GA1UdDgQWBBSFrBrRQ/fI\nrFXUxR1BSKvVeErUUzAPBgNVHRMBAf8EBTADAQH/MDoGA1UdHwQzMDEwL6AtoCuG\nKWh0dHBzOi8va2RzaW50Zi5hbWQuY29tL3ZjZWsvdjEvTWlsYW4vY3JsMEYGCSqG\nSIb3DQEBCjA5oA8wDQYJYIZIAWUDBAICBQChHDAaBgkqhkiG9w0BAQgwDQYJYIZI\nAWUDBAICBQCiAwIBMKMDAgEBA4ICAQC6m0kDp6zv4Ojfgy+zleehsx6ol0ocgVel\nETobpx+EuCsqVFRPK1jZ1sp/lyd9+0fQ0r66n7kagRk4Ca39g66WGTJMeJdqYriw\nSTjjDCKVPSesWXYPVAyDhmP5n2v+BYipZWhpvqpaiO+EGK5IBP+578QeW/sSokrK\ndHaLAxG2LhZxj9aF73fqC7OAJZ5aPonw4RE299FVarh1Tx2eT3wSgkDgutCTB1Yq\nzT5DuwvAe+co2CIVIzMDamYuSFjPN0BCgojl7V+bTou7dMsqIu/TW/rPCX9/EUcp\nKGKqPQ3P+N9r1hjEFY1plBg93t53OOo49GNI+V1zvXPLI6xIFVsh+mto2RtgEX/e\npmMKTNN6psW88qg7c1hTWtN6MbRuQ0vm+O+/2tKBF2h8THb94OvvHHoFDpbCELlq\nHnIYhxy0YKXGyaW1NjfULxrrmxVW4wcn5E8GddmvNa6yYm8scJagEi13mhGu4Jqh\n3QU3sf8iUSUr09xQDwHtOQUVIqx4maBZPBtSMf+qUDtjXSSq8lfWcd8bLr9mdsUn\nJZJ0+tuPMKmBnSH860llKk+VpVQsgqbzDIvOLvD6W1Umq25boxCYJ+TuBoa4s+HH\nCViAvgT9kf/rBq1d+ivj6skkHxuzcxbk1xv6ZGxrteJxVH7KlX7YRdZ6eARKwLe4\nAFZEAwoKCQ==\n-----END CERTIFICATE-----\n`
```
Comment (msanft) [inline internal/config/azure.go:21]: We could move this to a testdata package too, or make it readable within the file.

**Review by msanft — COMMENTED** [review]
Looked at the attestation-specific changes, which lgtm mostly

File: internal/config/azure.go (hunk 22 above)
Comment (derpsteb) [reply internal/config/azure.go:21]: I would like this to stay a constant as it is used during aTLS verification in the CLI. Overwriting this value on accident would be less than ideal.

File: internal/verify/verify_test.go [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+import (
+	"strings"
+	"testing"
+
+	"github.com/edgelesssys/constellation/v2/internal/logger"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseCerts(t *testing.T) {
+	validCert := `-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (derpsteb) [reply internal/verify/verify_test.go:17]: I moved all the values you commented on to a testdata package. But instead of embedding them I put the values into constants to prevent accidental overwrites. Let's talk in case you want to stick to our current approach.

File: internal/config/azure.go (hunk 22 above)
Comment (msanft) [reply internal/config/azure.go:21]: True, valid point.

**Review by elchead — APPROVED** [review]
could you run verify on AWS and include the upload step on your PR branch to test that everything works as expected? otherwise lgtm.

**Comment by derpsteb** [comment]
Verify [run](https://github.com/edgelesssys/constellation/actions/runs/6928948273) with patched condition.

**Comment by derpsteb** [comment]
had to rebase to use latest image in pipeline. didn't want to use outdated image.

**Review by daniel-weisse — APPROVED** [review]

**Comment by derpsteb** [comment]
Testrun was successful and API is serving SVNs that mitigate CacheWarp. I will wait for a comment from @thomasten and then revert the CI change (allowing non-main TCB value uploads), rebase, squash fixups and merge.

File: .github/actions/e2e_verify/action.yml [hunk 24]
```diff
@@ -66,7 +66,7 @@ runs:
           forwarderPID=$!
           sleep 5
 
-          if [[ ${{ inputs.cloudProvider }} == "azure" ]]; then
+          if [[ ${{ inputs.cloudProvider }} == "azure" || ${{ inputs.cloudProvider }} == "aws" ]]; then
             echo "Extracting Azure TCB versions for API update"
```
Comment (elchead) [inline .github/actions/e2e_verify/action.yml:70]:
```suggestion
            echo "Extracting TCB versions for API update"
```

File: internal/attestation/aws/snp/validator.go [hunk 25]
```diff
@@ -9,99 +9,216 @@ package snp
...
 	variant.AWSSEVSNP
+	// Embed validator to implement Validate method for aTLS handshake.
 	*vtpm.Validator
-	getDescribeClient func(context.Context, string) (awsMetadataAPI, error)
+	// AMD root key. Root of trust for the ASK used during report validation.
+	ark *x509.Certificate
+	// AMD signing key. Intermediate certificate that connects ARK and report signing key (leaf certificate).
+	ask *x509.Certificate
+	// cfg contains version numbers required for the SNP report validation.
+	cfg *config.AWSSEVSNP
```
Comment (thomasten) [inline internal/attestation/aws/snp/validator.go:43]: why do we duplicate ark and ask here if we also store the config?

File: internal/attestation/snp/snp.go [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+)
+
+// sevProduct is the product info of the SEV platform as reported through CPUID[EAX=1].
+// It may become necessary in the future to differentiate among CSP vendors.
+var sevProduct = &spb.SevProduct{Name: spb.SevProduct_SEV_PRODUCT_MILAN, Stepping: 0} // Milan-B0
+
+// Product returns the SEV product info currently supported by Constellation's SNP attestation.
+func Product() *spb.SevProduct {
+	return sevProduct
+}
```
Comment (thomasten) [inline internal/attestation/snp/snp.go:32]: can we move the global var into the func?

File: internal/config/azure.go (hunk 22 above)
Comment (thomasten) [inline internal/config/azure.go:21]: nit: doesn't need to be exported

Comment (thomasten) [inline internal/config/azure.go:21]: nit: as this is also used in aws.go, you may move it to attestation.go

**Review by thomasten — APPROVED** [review]
just some nits

File: internal/attestation/aws/snp/validator.go (hunk 25 above)
Comment (derpsteb) [reply internal/attestation/aws/snp/validator.go:43]: that does make no sense, right.

**Comment by derpsteb** [comment]
Final testrun with all suggestions and no temporary ci modifications: https://github.com/edgelesssys/constellation/actions/runs/6981570312
